File: euxis-crypto-packages/src/crypto-lib/src/core/service.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2024-2026 Euxis Contributors

"""Dependency injection architecture for crypto services."""
from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any


# Configuration class for crypto services
@dataclass
class CryptoConfig:
    """Configuration for crypto services."""

    default_encryption_algorithm: str
    default_hash_algorithm: str
    default_signature_algorithm: str
    key_rotation_interval: int
    enable_audit_logging: bool

# Algorithm registry interface for runtime algorithm selection
class AlgorithmRegistry(ABC):
    """Registry interface for cryptographic algorithms."""

    @abstractmethod
    def register_encryption_algorithm(self, name: str, algorithm: Any) -> None:
        """Register an encryption algorithm."""

    @abstractmethod
    def register_hash_algorithm(self, name: str, algorithm: Any) -> None:
        """Register a hash algorithm."""

    @abstractmethod
    def register_signature_algorithm(self, name: str, algorithm: Any) -> None:
        """Register a signature algorithm."""

    @abstractmethod
    def get_encryption_algorithm(self, name: str) -> Any | None:
        """Return an encryption algorithm by name."""

    @abstractmethod
    def get_hash_algorithm(self, name: str) -> Any | None:
        """Return a hash algorithm by name."""

    @abstractmethod
    def get_signature_algorithm(self, name: str) -> Any | None:
        """Return a signature algorithm by name."""

    @abstractmethod
    def list_available_algorithms(self) -> dict[str, list[str]]:
        """List all available algorithms by category."""

# Key management service interface
class KeyManagementService(ABC):
    """Key management service interface."""

    @abstractmethod
    async def generate_key(self) -> str:
        """Generate a new key."""

    @abstractmethod
    async def generate_key_pair(self) -> tuple[str, str]:
        """Generate a new key pair."""

    @abstractmethod
    async def rotate_key(self, key_id: str) -> str:
        """Rotate a key and return the new key ID."""

    @abstractmethod
    async def get_key(self, key_id: str) -> str | None:
        """Return a key by ID."""

    @abstractmethod
    async def revoke_key(self, key_id: str) -> None:
        """Revoke a key by ID."""

    @abstractmethod
    async def list_active_keys(self) -> list[str]:
        """List active key IDs."""

# Main crypto service with dependency injection
class CryptoService:
    """Crypto service orchestrating algorithms and keys."""

    def __init__(self,
                 config: CryptoConfig,
                 algorithm_registry: AlgorithmRegistry,
                 key_management_service: KeyManagementService) -> None:
        """Initialize the crypto service."""
        self.config = config
        self.algorithm_registry = algorithm_registry
        self.key_management_service = key_management_service

    async def encrypt(self, data: str, algorithm_name: str | None = None) -> dict[str, Any]:
        """Encrypt data using the configured algorithm."""
        algorithm = (algorithm_name if algorithm_name
                    else self.config.default_encryption_algorithm)

        if not self.algorithm_registry.get_encryption_algorithm(algorithm):
            msg = f"Encryption algorithm not found: {algorithm}"
            raise ValueError(msg)

        await self.key_management_service.generate_key()

        # Implementation would use actual crypto operations here
        return {
            "cipher_text": f"encrypted_{data}_{algorithm}",
            "algorithm": algorithm,
            "iv": "generated_iv",
            "auth_tag": "generated_auth_tag"
        }

    async def decrypt(self, cipher_text: str, algorithm_name: str | None = None) -> dict[str, Any]:
        """Decrypt data using the configured algorithm."""
        algorithm = (algorithm_name if algorithm_name
                    else self.config.default_encryption_algorithm)

        if not self.algorithm_registry.get_encryption_algorithm(algorithm):
            msg = f"Encryption algorithm not found: {algorithm}"
            raise ValueError(msg)

        # Implementation would use actual crypto operations here
        return {
            "plain_text": f"decrypted_{cipher_text}",
            "algorithm": algorithm
        }

    async def hash_data(self, input_data: str, algorithm_name: str | None = None) -> dict[str, Any]:
        """Hash data using the configured algorithm."""
        algorithm = (algorithm_name if algorithm_name
                    else self.config.default_hash_algorithm)

        if not self.algorithm_registry.get_hash_algorithm(algorithm):
            msg = f"Hash algorithm not found: {algorithm}"
            raise ValueError(msg)

        # Implementation would use actual hash operations here
        return {
            "hash": f"hash_{input_data}_{algorithm}",
            "algorithm": algorithm
        }

    async def sign(
        self,
        data: str,
        _private_key: str,
        algorithm_name: str | None = None,
    ) -> dict[str, Any]:
        """Sign data using the configured algorithm."""
        algorithm = (algorithm_name if algorithm_name
                    else self.config.default_signature_algorithm)

        if not self.algorithm_registry.get_signature_algorithm(algorithm):
            msg = f"Signature algorithm not found: {algorithm}"
            raise ValueError(msg)

        # Implementation would use actual signing operations here
        return {
            "signature": f"signature_{data}_{algorithm}",
            "algorithm": algorithm,
            "data": data
        }

    async def verify(
        self,
        signature: str,
        _public_key: str,
        algorithm_name: str | None = None,
    ) -> dict[str, Any]:
        """Verify a signature using the configured algorithm."""
        algorithm = (algorithm_name if algorithm_name
                    else self.config.default_signature_algorithm)

        if not self.algorithm_registry.get_signature_algorithm(algorithm):
            msg = f"Signature algorithm not found: {algorithm}"
            raise ValueError(msg)

        # Implementation would use actual verification here
        return {
            "is_valid": True,
            "signature": signature,
            "algorithm": algorithm
        }

    async def rotate_keys(self) -> None:
        """Rotate all active keys."""
        active_keys = await self.key_management_service.list_active_keys()
        for key_id in active_keys:
            await self.key_management_service.rotate_key(key_id)

    def get_available_algorithms(self) -> dict[str, list[str]]:
        """Return available algorithms by category."""
        return self.algorithm_registry.list_available_algorithms()

    def update_config(self, **kwargs: Any) -> None:
        """Update configuration with allowed values."""
        for key, value in kwargs.items():
            if hasattr(self.config, key):
                setattr(self.config, key, value)

File: euxis-crypto-packages/src/crypto-lib/src/core/service.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright (c) 2024-2026 Euxis Contributors
 */

// Dependency injection architecture for crypto services
import {
  SymmetricKey, PublicKey, PrivateKey, PlainText, CipherText, HashValue, Signature,
  EncryptionAlgorithm, HashAlgorithm, SignatureAlgorithm,
  EncryptionResult, DecryptionResult, HashResult, KeyPair,
  SignatureResult, VerificationResult,
  createHashValue, createSignature
} from '../types.js';

// Configuration interface for crypto services
export interface CryptoConfig {
  defaultEncryptionAlgorithm: EncryptionAlgorithm;
  defaultHashAlgorithm: HashAlgorithm;
  defaultSignatureAlgorithm: SignatureAlgorithm;
  keyRotationInterval: number;
  enableAuditLogging: boolean;
}

// Algorithm registry interface for runtime algorithm selection
export interface AlgorithmRegistry {
  registerEncryptionAlgorithm(name: string, algorithm: EncryptionAlgorithm): void;
  registerHashAlgorithm(name: string, algorithm: HashAlgorithm): void;
  registerSignatureAlgorithm(name: string, algorithm: SignatureAlgorithm): void;
  getEncryptionAlgorithm(name: string): EncryptionAlgorithm | undefined;
  getHashAlgorithm(name: string): HashAlgorithm | undefined;
  getSignatureAlgorithm(name: string): SignatureAlgorithm | undefined;
  listAvailableAlgorithms(): { encryption: string[], hash: string[], signature: string[] };
}

// Key management service interface
export interface KeyManagementService {
  generateKey(): Promise<SymmetricKey>;
  generateKeyPair(): Promise<KeyPair>;
  rotateKey(keyId: string): Promise<SymmetricKey>;
  getKey(keyId: string): Promise<SymmetricKey | undefined>;
  revokeKey(keyId: string): Promise<void>;
  listActiveKeys(): Promise<string[]>;
}

// Main crypto service with dependency injection
export class CryptoService {
  private config: CryptoConfig;
  private algorithmRegistry: AlgorithmRegistry;
  private keyManagementService: KeyManagementService;

  constructor(
    config: CryptoConfig,
    algorithmRegistry: AlgorithmRegistry,
    keyManagementService: KeyManagementService
  ) {
    this.config = config;
    this.algorithmRegistry = algorithmRegistry;
    this.keyManagementService = keyManagementService;
  }

  // Encrypt with configurable algorithm selection
  async encrypt(data: PlainText, algorithmName?: string): Promise<EncryptionResult> {
    const algorithm = algorithmName
      ? this.algorithmRegistry.getEncryptionAlgorithm(algorithmName)
      : this.config.defaultEncryptionAlgorithm;

    if (!algorithm) {
      throw new Error(`Encryption algorithm not found: ${algorithmName}`);
    }

    const key = await this.keyManagementService.generateKey();

    // Implementation would use actual crypto operations here
    return {
      cipherText: `encrypted_${data}_${algorithm}` as CipherText,
      algorithm,
      iv: 'generated_iv',
      authTag: 'generated_auth_tag'
    };
  }

  // Decrypt with algorithm detection
  async decrypt(cipherText: CipherText, algorithmName?: string): Promise<DecryptionResult> {
    const algorithm = algorithmName
      ? this.algorithmRegistry.getEncryptionAlgorithm(algorithmName)
      : this.config.defaultEncryptionAlgorithm;

    if (!algorithm) {
      throw new Error(`Encryption algorithm not found: ${algorithmName}`);
    }

    // Implementation would use actual crypto operations here
    return {
      plainText: `decrypted_${cipherText}` as PlainText,
      algorithm
    };
  }

  // Hash with configurable algorithm selection
  async hash(input: PlainText, algorithmName?: string): Promise<HashResult> {
    const algorithm = algorithmName
      ? this.algorithmRegistry.getHashAlgorithm(algorithmName)
      : this.config.defaultHashAlgorithm;

    if (!algorithm) {
      throw new Error(`Hash algorithm not found: ${algorithmName}`);
    }

    // Implementation would use actual hash operations here
    return {
      hash: createHashValue(`hash_${input}_${algorithm}`),
      algorithm
    };
  }

  // Sign with configurable algorithm selection
  async sign(data: PlainText, privateKey: PrivateKey, algorithmName?: string): Promise<SignatureResult> {
    const algorithm = algorithmName
      ? this.algorithmRegistry.getSignatureAlgorithm(algorithmName)
      : this.config.defaultSignatureAlgorithm;

    if (!algorithm) {
      throw new Error(`Signature algorithm not found: ${algorithmName}`);
    }

    // Implementation would use actual signing operations here
    return {
      signature: createSignature(`signature_${data}_${algorithm}`),
      algorithm,
      data
    };
  }

  // Verify signature with algorithm detection
  async verify(signature: Signature, publicKey: PublicKey, algorithmName?: string): Promise<VerificationResult> {
    const algorithm = algorithmName
      ? this.algorithmRegistry.getSignatureAlgorithm(algorithmName)
      : this.config.defaultSignatureAlgorithm;

    if (!algorithm) {
      throw new Error(`Signature algorithm not found: ${algorithmName}`);
    }

    // Implementation would use actual verification here
    return {
      isValid: true,
      signature,
      algorithm
    };
  }

  // Key management operations
  async rotateKeys(): Promise<void> {
    const activeKeys = await this.keyManagementService.listActiveKeys();
    for (const keyId of activeKeys) {
      await this.keyManagementService.rotateKey(keyId);
    }
  }

  // Get available algorithms for runtime selection
  getAvailableAlgorithms(): { encryption: string[], hash: string[], signature: string[] } {
    return this.algorithmRegistry.listAvailableAlgorithms();
  }

  // Update configuration at runtime
  updateConfig(newConfig: Partial<CryptoConfig>): void {
    this.config = { ...this.config, ...newConfig };
  }
}